### PR TITLE
[DB SCHEMA CHANGE] Dynamic tracking notifications

### PR DESCRIPTION
**Pull request by diegocepedaw**
Allow for providing tracking notification destinations on incident creation time

Create a new dynamic_tracking_notification table to allow piggybacking off of the existing template/plan UI and still allow arbitrary tracking notifications to be specified on incident creation time. This will work similarly to how dynamic plans work today where the targets passed at incident creation time and are stored in a dynamic_plan_map table ato be later used to render the appropriate messages to the ad-hoc targets. 

To leverage this new feature a client would pass at incident creation, in addition to the context, a dynamic_tracking_notifications json field with the type of message and destinations for their dynamic tracking notifications. If a plan has dynamic_tracking enabled then the message processor will ignore the plan defined tracking notification and reuse the tracking template defined in the plan with the targets provided in dynamic_tracking_notifications. If the processor can’t find an appropriate template it will create an error message to expose the rendering failure to the end user like it currently exposes other messages that fail to send or render. In order to make this more explicit we will also add an extra `dynamic_tracking` tinyint(1) NOT NULL DEFAULT '0' column to the plan table and a checkbox in the plan editing UI so users can specify if they want to enable dynamic tracking messages for that plan

The contents of the dynamic_tracking_notifications table will be purged when the corresponding incidents are purged for retention.

### CODE REVIEW

File: src/iris/api.py [hunk 1]
```diff
@@ -1999,13 +2016,54 @@ def on_post(self, req, resp):
...
+
+        dynamic_tracking_notifications = incident_params.get(
+            "dynamic_tracking_notifications", []
+        )
+        if not isinstance(dynamic_tracking_notifications, list):
+            raise falcon.HTTPBadRequest(
+                "Invalid Format", "dynamic_tracking_notifications should be a list."
+            )
+        for notification in dynamic_tracking_notifications:
+            if not isinstance(notification, dict):
```
Comment (jrgp) [inline src/iris/api.py:2046]: At some point, using a library like marshmallow would help simplify this greatly by avoiding all of these manual checks

File: src/iris/api.py [hunk 2]
```diff
@@ -2269,9 +2360,17 @@ def on_get(self, req, resp, incident_id):
 
             cursor.execute(single_incident_query_tags, incident['id'])
             incident['tags'] = cursor.fetchall()
-            connection.close()
 
             incident['context'] = ujson.loads(incident['context'])
+            # retrieve dynamic_tracking_notification for each incident
+            incident['dynamic_tracking'] = []
```
Comment (jrgp) [inline src/iris/api.py]: Could this just be assigned to cursor.fetchall() like in the above examples, avoiding the for loop/etc?

File: src/iris/api.py [hunk 3]
```diff
@@ -1179,6 +1187,7 @@ def on_get(self, req, resp, plan_id):
 
             cursor.execute(single_plan_query_tags, plan['id'])
             plan['tags'] = cursor.fetchall()
+            plan['dynamic_tracking'] = bool(plan.get('dynamic_tracking'))
```
Comment (jrgp) [inline src/iris/api.py]: Nit: could we use [] instead of get() here, as the field is guaranteed to be in the response as it is in the select query?

Comment (diegocepedaw) [reply src/iris/api.py]: changed

File: src/iris/api.py (hunk 2 above)
Comment (diegocepedaw) [reply src/iris/api.py]: changed

File: src/iris/api.py (hunk 1 above)
Comment (diegocepedaw) [reply src/iris/api.py:2046]: yeah, we could definitely benefit massively from adding that style of validation to all endpoints

**Review by jrgp — APPROVED** [review]

**Pull request merged**